File: ts/packages/katari-runtime/src/runtime/for-loop.ts
import type { AgentState, Event, Signal } from "./types.js";
import {
  getVar, setVar, findFor, findThread,
  finishThread, resumeThread, spawnChildThread,
} from "./types.js";

// ===========================================================================
// Setup
// ===========================================================================

export function setupFor(
  agent: AgentState,
  threadId: number,
  dst: number,
  fid: number,
  events: Event[]
): void {
  const fdef = findFor(agent.module, fid)!;

  // Initialize state vars
  for (let i = 0; i < fdef.stateVars.length; i++) {
    setVar(agent, fdef.stateVars[i]!, getVar(agent, fdef.stateInits[i]!));
  }

  // Calculate min array length
  let minLen = Infinity;
  for (const arrVar of fdef.arrays) {
    const arr = getVar(agent, arrVar);
    minLen = Math.min(minLen, Array.isArray(arr) ? arr.length : 0);
  }
  if (!isFinite(minLen)) minLen = 0;

  const t = agent.threads.get(threadId)!;
  t.status = {
    tag: "Suspended",
    reason: { tag: "For", forDefId: fid, currentIndex: 0, minLength: minLen, dst },
  };

  if (minLen > 0) {
    startForIteration(agent, threadId, fid, 0, events);
  } else {
    finishForLoop(agent, threadId, fid, dst, events);
  }
}

// ===========================================================================
// Iteration
// ===========================================================================

function startForIteration(
  agent: AgentState,
  parentThreadId: number,
  fid: number,
  index: number,
  events: Event[]
): void {
  const fdef = findFor(agent.module, fid)!;
  for (let i = 0; i < fdef.iterVars.length; i++) {
    const arr = getVar(agent, fdef.arrays[i]!);
    if (Array.isArray(arr)) {
      setVar(agent, fdef.iterVars[i]!, arr[index] ?? null);
    }
  }
  spawnChildThread(agent, fdef.body, "ForBody", parentThreadId, events);
}

function advanceFor(
  agent: AgentState,
  ownerThreadId: number,
  fid: number,
  currentIndex: number,
  minLength: number,
  dst: number,
  events: Event[]
): void {
  const nextIndex = currentIndex + 1;
  if (nextIndex < minLength) {
    const t = agent.threads.get(ownerThreadId)!;
    if (t.status.tag === "Suspended" && t.status.reason.tag === "For") {
      t.status.reason.currentIndex = nextIndex;
    }
    startForIteration(agent, ownerThreadId, fid, nextIndex, events);
  } else {
    finishForLoop(agent, ownerThreadId, fid, dst, events);
  }
}

function finishForLoop(
  agent: AgentState,
  ownerThreadId: number,
  fid: number,
  dst: number,
  events: Event[]
): void {
  const fdef = findFor(agent.module, fid)!;
  if (fdef.then !== null) {
    spawnChildThread(agent, fdef.then, "ForThen", ownerThreadId, events);
  } else {
    setVar(agent, dst, null);
    resumeThread(agent, ownerThreadId, events);
  }
}

// ===========================================================================
// Signals
// ===========================================================================

export function processForBodySignal(
  agent: AgentState,
  ownerThreadId: number,
  signal: Signal,
  events: Event[]
): void {
  const t = agent.threads.get(ownerThreadId);
  if (!t || t.status.tag !== "Suspended" || t.status.reason.tag !== "For") return;
  const reason = t.status.reason;

  switch (signal.tag) {
    case "ForContinue":
      for (const [sv, nv] of signal.mutations) setVar(agent, sv, getVar(agent, nv));
      advanceFor(agent, ownerThreadId, reason.forDefId, reason.currentIndex, reason.minLength, reason.dst, events);
      break;
    case "Normal":
      advanceFor(agent, ownerThreadId, reason.forDefId, reason.currentIndex, reason.minLength, reason.dst, events);
      break;
    case "ForBreak":
      setVar(agent, reason.dst, signal.value);
      resumeThread(agent, ownerThreadId, events);
      break;
    case "FnReturn":
    case "HandleBreak":
      finishThread(agent, ownerThreadId, signal, events);
      break;
    default:
      break;
  }
}

export function processForThenSignal(
  agent: AgentState,
  ownerThreadId: number,
  signal: Signal,
  events: Event[]
): void {
  const t = agent.threads.get(ownerThreadId);
  if (!t || t.status.tag !== "Suspended" || t.status.reason.tag !== "For") return;

  switch (signal.tag) {
    case "Normal":
      setVar(agent, t.status.reason.dst, signal.value);
      resumeThread(agent, ownerThreadId, events);
      break;
    case "FnReturn":
      finishThread(agent, ownerThreadId, signal, events);
      break;
    default:
      break;
  }
}

File: rust/katari-runtime/src/runtime/for_loop.rs
use crate::ir::IRModule;
use crate::value::Value;

use super::agent::AgentState;
use super::event::Event;
use super::execute::{finish_thread, resume_thread, spawn_child_thread};
use super::signal::Signal;
use super::thread::{SuspendReason, ThreadStatus};

/// Execute For instruction: init state, suspend parent, start first iteration.
pub fn setup(
    agent: &mut AgentState,
    module: &IRModule,
    thread_id: u32,
    dst: u32,
    fid: u32,
    events: &mut Vec<Event>,
) {
    let for_def = module
        .fors
        .iter()
        .find(|f| f.id == fid)
        .expect("for def not found");

    // Initialize state variables
    for (sv, iv) in for_def
        .state_vars
        .iter()
        .zip(for_def.state_inits.iter())
    {
        let val = agent.get_var(*iv);
        agent.set_var(*sv, val);
    }

    // Calculate min array length
    let min_len = for_def
        .arrays
        .iter()
        .map(|arr_var| {
            agent
                .get_var(*arr_var)
                .as_array()
                .map(|a| a.len())
                .unwrap_or(0)
        })
        .min()
        .unwrap_or(0) as u32;

    // Suspend parent
    if let Some(t) = agent.threads.get_mut(&thread_id) {
        t.status = ThreadStatus::Suspended(SuspendReason::For {
            for_def_id: fid,
            current_index: 0,
            min_length: min_len,
            dst,
        });
    }

    if min_len > 0 {
        start_iteration(agent, module, thread_id, fid, 0, events);
    } else {
        finish_loop(agent, module, thread_id, fid, dst, events);
    }
}

/// Start a single iteration of the for loop.
fn start_iteration(
    agent: &mut AgentState,
    module: &IRModule,
    parent_thread_id: u32,
    fid: u32,
    index: u32,
    events: &mut Vec<Event>,
) {
    let for_def = module
        .fors
        .iter()
        .find(|f| f.id == fid)
        .expect("for def not found");

    // Bind element variables
    for (i, iter_var) in for_def.iter_vars.iter().enumerate() {
        if let Some(arr) = agent.get_var(for_def.arrays[i]).as_array() {
            if let Some(elem) = arr.get(index as usize) {
                agent.set_var(*iter_var, elem.clone());
            }
        }
    }

    spawn_child_thread(
        agent,
        for_def.body,
        crate::ir::ThreadKind::ForBody,
        parent_thread_id,
        events,
    );
}

/// Process signal from a completed ForBody thread.
pub fn process_body_signal(
    agent: &mut AgentState,
    module: &IRModule,
    owner_thread_id: u32,
    signal: Signal,
    events: &mut Vec<Event>,
) {
    let (fid, current_index, min_length, dst) = match agent.threads.get(&owner_thread_id) {
        Some(t) => match &t.status {
            ThreadStatus::Suspended(SuspendReason::For {
                for_def_id,
                current_index,
                min_length,
                dst,
            }) => (*for_def_id, *current_index, *min_length, *dst),
            _ => return,
        },
        None => return,
    };

    match signal {
        Signal::ForContinue(mutations) => {
            for (sv, nv) in &mutations {
                let val = agent.get_var(*nv);
                agent.set_var(*sv, val);
            }
            advance(
                agent,
                module,
                owner_thread_id,
                fid,
                current_index,
                min_length,
                dst,
                events,
            );
        }
        Signal::Normal(_) => {
            // Treat as ForContinue([]) — no mutations
            advance(
                agent,
                module,
                owner_thread_id,
                fid,
                current_index,
                min_length,
                dst,
                events,
            );
        }
        Signal::ForBreak(value) => {
            agent.set_var(dst, value);
            resume_thread(agent, owner_thread_id, events);
        }
        Signal::FnReturn(value) => {
            finish_thread(agent, owner_thread_id, Signal::FnReturn(value), events);
        }
        Signal::HandleBreak(value) => {
            finish_thread(
                agent,
                owner_thread_id,
                Signal::HandleBreak(value),
                events,
            );
        }
        _ => {}
    }
}

/// Advance to the next iteration or finish.
fn advance(
    agent: &mut AgentState,
    module: &IRModule,
    owner_thread_id: u32,
    fid: u32,
    current_index: u32,
    min_length: u32,
    dst: u32,
    events: &mut Vec<Event>,
) {
    let next_index = current_index + 1;
    if next_index < min_length {
        if let Some(t) = agent.threads.get_mut(&owner_thread_id) {
            t.status = ThreadStatus::Suspended(SuspendReason::For {
                for_def_id: fid,
                current_index: next_index,
                min_length,
                dst,
            });
        }
        start_iteration(agent, module, owner_thread_id, fid, next_index, events);
    } else {
        finish_loop(agent, module, owner_thread_id, fid, dst, events);
    }
}

/// Finish for loop (all iterations done or empty array).
fn finish_loop(
    agent: &mut AgentState,
    module: &IRModule,
    owner_thread_id: u32,
    fid: u32,
    dst: u32,
    events: &mut Vec<Event>,
) {
    let for_def = module
        .fors
        .iter()
        .find(|f| f.id == fid)
        .expect("for def not found");

    if let Some(then_tid) = for_def.then {
        spawn_child_thread(
            agent,
            then_tid,
            crate::ir::ThreadKind::ForThen,
            owner_thread_id,
            events,
        );
    } else {
        agent.set_var(dst, Value::Null);
        resume_thread(agent, owner_thread_id, events);
    }
}

/// Process signal from a completed ForThen thread.
pub fn process_then_signal(
    agent: &mut AgentState,
    owner_thread_id: u32,
    signal: Signal,
    events: &mut Vec<Event>,
) {
    let dst = match agent.threads.get(&owner_thread_id) {
        Some(t) => match &t.status {
            ThreadStatus::Suspended(SuspendReason::For { dst, .. }) => *dst,
            _ => return,
        },
        None => return,
    };

    match signal {
        Signal::Normal(value) => {
            agent.set_var(dst, value);
            resume_thread(agent, owner_thread_id, events);
        }
        Signal::FnReturn(value) => {
            finish_thread(agent, owner_thread_id, Signal::FnReturn(value), events);
        }
        _ => {}
    }
}
